docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,147 @@
+
+Because ASO already uses the new APIs, adopting ASO in CAPZ accomplishes the goal of migrating off of the deprecated APIs. In addition, the declarative, Kubernetes-native interface exposed by ASO unlocks opportunities to simplify how CAPZ reconciles resources in Azure.
+
+### Goals
+
+- Adopt ASO in CAPZ to handle creating, updating, and deleting Azure resources
+- Determine a method to migrate CAPZ's individual Azure service interfaces one-by-one to use ASO
+- Determine how ASO will be installed on Cluster API management clusters
+
+### Non-Goals/Future Work
+
+- Entirely eliminate direct usage of any Azure Go SDK
+- Handle changes to Azure resources asynchronously
+- Allow users to interact with ASO resources directly, out-of-band with CAPZ
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+TODO
+
+### Requirements (Optional)
+
+#### Functional Requirements
+
+##### FR1
+
+TODO
+
+#### Non-Functional Requirements
+
+##### NFR1
+
+TODO
+
+### Implementation Details/Notes/Constraints
+
+- ASO needs to be installed on the management cluster
+  - Multi- or single-tenant?
+  - Install in capz-controller-manager `func main()`
+- GET requests are not as straightforward in the ASO model because BYO resources
+  may not initially exist in ASO
+- Current service pattern in CAPZ works for Azure services that are not
+  represented in ASO (resource health)
+- Each release of ASO can expose several Azure API versions of each resource,
+  including GA and preview API versions.
+
+### Security Model
+
+- ASO resources are subject to Kubernetes RBAC => capz-controller-manager will require many more permissions that may extend beyond what's minimally required.
+  - If a user already has their own ASO resources installed, CAPZ may require
+    permission to operate against them to handle the ASO resources it needs to
+    manage.
+
+### Risks and Mitigations
+
+- Introduction of functional regressions
+- Increased management cluster API server traffic
+- Conflicting user installations of ASO or ASO resources
+- Future breaking changes in ASO
+
+## Alternatives
+
+- Adopt new SDK directly
+  - Likely much more straightforward transition => can be done faster with lower risk
+
+## Upgrade Strategy
+
+- There will be at least one upgrade path between CAPZ versions where ASO resources don't exist in the old version and they're expected to exist in the new version
+  - Need a one-time bootstrapping process to create managed resources in ASO?
+- Once ASO is integrated, updates to CAPZ may include updates to the API versions used for ASO resources.
+  - API versions of ASO resources may also be periodically removed.
+
+## Additional Details
+
+### Test Plan
+
+- Leverage existing e2e tests
+- Add unit tests for new ASO integration
+- Run one-off tests against large clusters to catch performance regressions
+
+### Graduation Criteria
+
+TODO
```

Comment:
It looks like this section is intended for discrete features that matriculate from alpha -> beta -> GA. From our discussions so far I wasn't sure if were planning to have a similar process for these changes. Should we be planning to do something like this for ASO integration?

cc @CecileRobertMichon

Reply:
No I don't think so, ASO is an implementation detail, not a new feature under a feature flag 

If anything this can talk about the "rollout" process for all the services and how we're planning to switch them over one by one